packages/preferences/src/browser/preferences-tree-widget.ts

```diff
@@ -0,0 +1,307 @@
+/********************************************************************************
+ * Copyright (C) 2018 Red Hat, Inc. and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { inject } from "inversify";
+import { Message } from "@phosphor/messaging";
+import { PreferencesMenuFactory } from "./preferences-menu-factory";
+import { PreferencesDecorator } from "./preferences-decorator";
+import { toArray } from '@phosphor/algorithm';
+import { DockPanel } from "@phosphor/widgets";
+import {
+    ContextMenuRenderer,
+    ExpandableTreeNode,
+    PreferenceSchemaProvider,
+    SelectableTreeNode,
+    PreferenceProperty,
+    PreferenceScope,
+    PreferenceService,
+    TreeModel,
+    TreeNode,
+    TreeProps,
+    TreeWidget,
+    WidgetManager,
+    Saveable,
+    ApplicationShell
+} from "@theia/core/lib/browser";
+import { UserPreferenceProvider } from "./user-preference-provider";
+import { WorkspacePreferenceProvider } from "./workspace-preference-provider";
+import { EditorBasedSplitPanel, EditorWidget } from "@theia/editor/lib/browser";
+import { DisposableCollection, Emitter, Event, MaybePromise, MessageService } from '@theia/core';
+import { PREFERENCES_CONTAINER_WIDGET_ID, PREFERENCES_TREE_WIDGET_ID } from "./preferences-contribution";
+
+let dirty = false;
+let currentEditor: EditorWidget;
+let preferencesScope: PreferenceScope;
+const onDirtyChangedEmitter = new Emitter<void>();
```

Comment:
please check the life-cycle. it looks like, this will leak.

also, please move to a `namespace PreferencesContainer` if it's unrelated, or add to class if, it's a property.

Reply:
fixed